File: customer-address-jpa/src/main/java/com/izertis/example/domain/PaymentMethod.java
package com.izertis.example.domain;

import jakarta.persistence.*;
import jakarta.validation.constraints.NotNull;
import org.hibernate.annotations.Cache;
import org.hibernate.annotations.CacheConcurrencyStrategy;
import org.springframework.data.annotation.CreatedBy;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedBy;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import java.io.Serializable;
import java.time.LocalDateTime;

/** */
@lombok.Getter
@lombok.Setter
@Entity
@Table(name = "payment_method")
@Cache(usage = CacheConcurrencyStrategy.READ_WRITE)
@EntityListeners(AuditingEntityListener.class)
public class PaymentMethod implements Serializable {

    @java.io.Serial
    private static final long serialVersionUID = 1L;

    @Id
    @GeneratedValue(strategy = GenerationType.SEQUENCE)
    @SequenceGenerator(name = "sequenceGenerator")
    private Long id;

    @Version
    private Integer version;

    @NotNull
    @Column(name = "type", nullable = false)
    @Convert(converter = PaymentMethodType.PaymentMethodTypeConverter.class)
    private PaymentMethodType type;

    @NotNull
    @Column(name = "card_number", nullable = false)
    private String cardNumber;

    @NotNull
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "customer_id")
    private Customer customer;

    @CreatedBy
    @Column(name = "created_by", updatable = false)
    protected String createdBy;

    @CreatedDate
    @Column(name = "created_date", columnDefinition = "TIMESTAMP", updatable = false)
    protected LocalDateTime createdDate;

    @LastModifiedBy
    @Column(name = "last_modified_by")
    protected String lastModifiedBy;

    @LastModifiedDate
    @Column(name = "last_modified_date", columnDefinition = "TIMESTAMP")
    protected LocalDateTime lastModifiedDate;

    /*
     * https://vladmihalcea.com/the-best-way-to-implement-equals-hashcode-and-tostring-
     * with-jpa-and-hibernate/
     */
    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (!(o instanceof PaymentMethod)) {
            return false;
        }
        PaymentMethod other = (PaymentMethod) o;
        return getId() != null && getId().equals(other.getId());
    }

    @Override
    public int hashCode() {
        return getClass().hashCode();
    }

}

File: kustomer-address-jpa/src/main/kotlin/com/izertis/example/domain/PaymentMethod.kt
package com.izertis.example.domain

import jakarta.persistence.*
import jakarta.validation.constraints.NotNull
import org.hibernate.annotations.Cache
import org.hibernate.annotations.CacheConcurrencyStrategy
import org.springframework.data.annotation.CreatedBy
import org.springframework.data.annotation.CreatedDate
import org.springframework.data.annotation.LastModifiedBy
import org.springframework.data.annotation.LastModifiedDate
import org.springframework.data.jpa.domain.support.AuditingEntityListener
import java.io.Serializable
import java.time.LocalDateTime

/**  */
@Entity
@Table(name = "payment_method")
@Cache(usage = CacheConcurrencyStrategy.READ_WRITE)
@EntityListeners(AuditingEntityListener::class)
data class PaymentMethod(
    @Id @GeneratedValue(strategy = GenerationType.SEQUENCE) var id: Long? = null,
    @Version var version: Int? = null,
    @NotNull
    @Column(name = "type", nullable = false)
    @Convert(converter = PaymentMethodType.PaymentMethodTypeConverter::class)
    var type: PaymentMethodType? = null,
    @NotNull @Column(name = "card_number", nullable = false) var cardNumber: String? = null,
    @NotNull
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "customer_id")
    @com.fasterxml.jackson.annotation.JsonBackReference
    var customer: Customer? = null
) : Serializable {

  companion object {
    private const val serialVersionUID = 1L
  }

  @CreatedBy @Column(name = "created_by", updatable = false) var createdBy: String? = null

  @CreatedDate
  @Column(name = "created_date", columnDefinition = "TIMESTAMP", updatable = false)
  var createdDate: LocalDateTime? = null

  @LastModifiedBy @Column(name = "last_modified_by") var lastModifiedBy: String? = null

  @LastModifiedDate
  @Column(name = "last_modified_date", columnDefinition = "TIMESTAMP")
  var lastModifiedDate: LocalDateTime? = null

  override fun toString(): String {
    return this::class.java.name + "#" + id
  }

  /* https://vladmihalcea.com/the-best-way-to-implement-equals-hashcode-and-tostring-with-jpa-and-hibernate/ */
  override fun equals(other: Any?): Boolean {
    if (this === other) return true
    if (other !is PaymentMethod) return false
    return id != null && id == other.id
  }

  override fun hashCode(): Int {
    return javaClass.hashCode()
  }
}
